syntax = "proto3";

package veritas.auth.v1;

import "google/protobuf/timestamp.proto";

message SrpParameters {
  string group = 1;
  string hash = 2;
}

message BeginAuthRequest {
  string login_username = 1;
  bytes client_start = 2;
}

message BeginAuthResponse {
  bytes salt = 1;
  bytes server_public = 2;
  string session_id = 3;
  SrpParameters params = 4;
}

message FinishAuthRequest {
  string session_id = 1;
  bytes client_proof = 2;
}

message FinishAuthResponse {
  bytes server_proof = 1;
  string user_uuid = 2;
  bytes refresh_token = 3;
  google.protobuf.Timestamp expires_at = 4;
}

message RevokeTokenRequest {
  bytes refresh_token = 1;
  string reason = 2;
}

message RevokeTokenResponse {
  bool revoked = 1;
}

enum TokenStatusState {
  TOKEN_STATUS_STATE_UNSPECIFIED = 0;
  TOKEN_STATUS_STATE_ACTIVE = 1;
  TOKEN_STATUS_STATE_REVOKED = 2;
  TOKEN_STATUS_STATE_UNKNOWN = 3;
}

message GetTokenStatusRequest {
  bytes refresh_token = 1;
}

message GetTokenStatusResponse {
  TokenStatusState state = 1;
  string reason = 2;
  google.protobuf.Timestamp revoked_at = 3;
  string user_uuid = 4;
}

service Gatekeeper {
  rpc BeginAuth(BeginAuthRequest) returns (BeginAuthResponse);
  rpc FinishAuth(FinishAuthRequest) returns (FinishAuthResponse);
  rpc RevokeToken(RevokeTokenRequest) returns (RevokeTokenResponse);
  rpc GetTokenStatus(GetTokenStatusRequest) returns (GetTokenStatusResponse);
}
